### ISSUE TITLE
Reading tetrahedral meshes (.msh/.vtk file formats) and feed to graph Nets

### ISSUE THREAD

**Issue by isurusuranga** [pyg-team/pytorch_geometric#1574]
## ❓ Questions & Help

I'm new to pytorch-geometric and I am trying to feed 3D volumetric meshes (i.e. .msh) into graph neural nets. I have gone though the documentation and I have found that pytorch-geometric supports trimesh. But I'm not sure that trimesh can load .msh files properly means may be some data get lost on reload.
I have produced volumetric meshes using gmsh and now I need to feed these data into graph neural nets using the pytorch-geometric. Is there a way to feed this type of data ? I really appreciate your help in this matter. I have attached a sample .msh file for your further reference. 

[ellipsoid_0.zip](https://github.com/rusty1s/pytorch_geometric/files/5125649/ellipsoid_0.zip)

**Comment by rusty1s** [comment]
Hi and thanks for this issue. Loading `*.msh` files into PyG is currently not supported, I am sorry! You may need to write your own processing procedure to convert these types of meshes into a graph-structured form. If you are interested in doing so, please consider sending a PR. I would be happy to help you out :)

**Comment by isurusuranga** [comment]
I just implemented two functions as below using the meshio package (pip3 install meshio). This is something very similar to the trimesh functionalities in the util package in pytorch geometric.

```
def from_meshio(mesh):
    r"""Converts a :.msh file to a
    :class:`torch_geometric.data.Data` instance.

    Args:
        mesh (meshio.read): A :obj:`meshio` mesh.
    """

    if meshio is None:
        raise ImportError('Package `meshio` could not be found.')

    pos = torch.from_numpy(mesh.points).to(torch.float)
    face = torch.from_numpy(mesh.cells[0].data).to(torch.long).t().contiguous()
    tetra = torch.from_numpy(mesh.cells[1].data)

    return Data(pos=pos, face=face, tetra=tetra)

def to_meshio(data):
    r"""Converts a :class:`torch_geometric.data.Data` instance to a
    :obj:`.msh format`.

    Args:
        data (torch_geometric.data.Data): The data object.
    """

    if meshio is None:
        raise ImportError('Package `meshio` could not be found.')
    
    points = data.pos.detach().cpu().numpy()
    faces = data.face.detach().t().cpu().numpy()
    tetra = data.tetra.detach().cpu().numpy()
    
    cells = [("triangle", faces), ("tetra", tetra)]

    return meshio.Mesh(points, cells)
```

In meshio, they currently support gmsh 2.2 version for writing a mesh object to a .msh file and there is a open issue for gmsh 4.1.
(https://github.com/nschloe/meshio/issues/865)

Therefore, if we need to write an output mesh object to the disk after we invoking to_meshio(), then we can use below line of code for that.
meshio.write('./test_results/test_mesh.msh', deformed_mesh, file_format="gmsh22") 

Again, if we need to visualze the output, we can directly use gmsh toolkit UI (https://gmsh.info/). In order to convert this mesh object to the latest gmsh version, we can simply run below command in a cmd after installing the gmsh-sdk package (pip3 install gmsh-sdk)
>gmsh -convert ./test_results/test_mesh.msh

code example of usage.

```
mesh = meshio.read(mesh_path)
data = from_meshio(mesh)
data
```
>Data(face=[3, 4704], pos=[4883, 3], tetra=[22687, 4])

```
predicted_mesh = to_meshio(data)
meshio.write('./out.msh', predicted_mesh, file_format="gmsh22")
```

**Comment by rusty1s** [comment]
That looks great! Please feel free to send a PR in case you are interested.

**Comment by isurusuranga** [comment]
I think I have mistakenly mentioned a few things wrong here. The volumetric meshes that I have generated consists of tetrahedrons. A tetrahedron is something that looks like the attached image.

![tetrahedron_img](https://user-images.githubusercontent.com/1329800/92932421-cc3d3580-f43c-11ea-90ad-6ddf4a544fd9.png)

Here, each row indicates the vertex indices that are associated with each tetrahedron (see the output of ellispoid.cells[0].data below). 

```
// load volumetric mesh object
msh_file = './ellispoid.vtk'
ellispoid = meshio.read(msh_file)
ellispoid 
```

> <meshio mesh object>
>   Number of points: 286
>   Number of cells:
>     tetra: 927
>   Cell data: CellEntityIds

`ellispoid.cells[0].data`

> array([[  0,   1,   2,   3],
>        [  4,   5,   6,   7],
>        [  0,   8,   1,   3],
>        ...,
>        [ 19, 142, 135,  37],
>        [ 27, 127, 172, 130],
>        [ 27, 172, 127,  79]], dtype=int64)


I have just applied changes to the aforementioned functionalities as below by simply removing the face attribute since we need to get all the points/nodes on the surface as well as in the interior.

```
def from_meshio(mesh):
    r"""Converts a :.msh file to a
    :class:`torch_geometric.data.Data` instance.

    Args:
        mesh (meshio.read): A :obj:`meshio` mesh.
    """

    if meshio is None:
        raise ImportError('Package `meshio` could not be found.')

    pos = torch.from_numpy(mesh.points).to(torch.float)
    tetra = torch.from_numpy(mesh.cells[0].data).to(torch.long).t().contiguous()

    return Data(pos=pos, tetra=tetra)


def to_meshio(data):
    r"""Converts a :class:`torch_geometric.data.Data` instance to a
    :obj:`.msh format`.

    Args:
        data (torch_geometric.data.Data): The data object.
    """

    if meshio is None:
        raise ImportError('Package `meshio` could not be found.')
    
    points = data.pos.detach().cpu().numpy()
    tetra = data.tetra.detach().t().cpu().numpy()
    
    cells = [("tetra", tetra)]

    return meshio.Mesh(points, cells) 

```

code example of usage.

```
mesh = meshio.read(mesh_path)
data = from_meshio(mesh)
data
```

> Data(pos=[286, 3], tetra=[4, 927])

I have introduced data.tetra to save the connectivity of tetras from a 3D volumetric mesh in a tensor with shape [4, num_tetras] and type torch.long.

**Now in order to transform this tetra to edge_index I should re-implement a class called TetraToEdge similar to FaceToEdge class.** 

**Can you please help to implement this line of code for the TetraToEdge ?**
`edge_index = torch.cat([face[:2], face[1:], face[::2]], dim=1) `

**Comment by isurusuranga** [comment]
I just resolved this.
`edge_index = torch.cat([tetra[:2], tetra[1:3, :], tetra[-2:], tetra[::2], tetra[::3], tetra[1::2]], dim=1)`

**Comment by rusty1s** [comment]
No, I don't think so :)

**Comment by javirk** [comment]
Hello, looking at the code and using my own examples, I wonder if the line 
`tetra = torch.from_numpy(mesh.cells[0].data).to(torch.long).t().contiguous()`
is correct. In my case mesh.cells[0]  doesn't have "tetra" information but vertex, wouldn't it be better to use
`tetra = torch.from_numpy(mesh.cells_dict['tetra']).to(torch.long).t().contiguous()`
instead?

In my case `(mesh.cells[1].data == mesh.cells_dict['tetra']).all()` returns `True`, so I guess they contain the same information, but I haven't found any documentation.

**Comment by rusty1s** [comment]
Thanks, can you post that information as well in the PR? https://github.com/rusty1s/pytorch_geometric/pull/2808

**Comment by isurusuranga** [comment]
Thank you for bringing this to my attention. I use a.vtk file to test this approach, and it contains a cells property in the 0th element, as seen below. In order to extract tetra, I utilised ref.cells[0].data.

[CellBlock(type='tetra', data=array([[172, 275, 279, 574],
[640, 700, 581, 767],
[459, 521, 479, 557],
...,
[787, 805, 719, 812],
[776, 787, 678, 812],
[776, 805, 787, 812]]))]

Anyway I didn't realize that it had an additional attribute called cells_dict['tetra']. Thank you for bringing this to my attention once more. I just checked it, and you're right, they have the same information. I just applied the modifications to my pull request.

Finally, I can confirm that it is consistent with tetrahedral meshes in .msh and .vtk formats after this modification. :)

I suppose the test failure is due to a ModuleNotFoundError i.e. meshio package.  How can I resolve that issue ?

**Comment by isurusuranga** [comment]
Some checks were not successful, which I suppose is due to a ModuleNotFoundError, i.e. the meshio package. What is the best way for me to tackle this problem ?

**Comment by rusty1s** [comment]
See https://github.com/rusty1s/pytorch_geometric/pull/2808#issuecomment-894025404

**Comment by gabyweg76** [comment]
I am interested to know how to reverse the operation, to pass from graph nets to .vtk file format : 
```
edge_index = torch.cat([tetra[:2], tetra[1:3, :], tetra[-2:], tetra[::2], tetra[::3], tetra[1::2]], dim=1) 
```
where 
```
edge_index.shape
> [2, ...] 
```
to 
```
tetra=[4, ...]
```
It seems we need to reverse the torch.cat operation but I don't know how to proceed.  Does anybody have a hint on this? This means a lot to me, thanks !!

**Comment by rusty1s** [comment]
I don‘t think this is a reversible operation. A graph is not necessarily a mesh. If applicable, try to keep holding both representations.

**Comment by gabyweg76** [comment]
Thank you for your answer. Yes, you're right, ...  My problem is to try, after partitioning a mesh with ClusterData, to do the reverse operation: to gather all the partitions' features data_part.x  into a great vector data.x (with the size of the supergraph features number) in the same order as the super graph before the partition. Do you have any clue to do that? Is there a corresponding system like the local coordinates system for each partitioning associated with the global coordinates of the super-graph? 

Actually, for a single graph without partition, I managed to do the following to refind the original tetra matrix:
```
a = torch.tensor_split(edge_index, 6, dim=1)
final_tetra = torch.zeros(4, a[0].shape[1])
final_tetra[0,:] =  a[0][0]
final_tetra[1,:] =  a[0][1]
final_tetra[2,:] =  a[1][1]
final_tetra[3,:] =  a[2][1]
```
But once we did a partition I didn't manage to find the same order for the tetra matrix as the original one...

**Comment by rusty1s** [comment]
I don't think something like that is possible with `ClusterLoader` - you would have to utilize an approach similar to [this](https://ps.is.mpg.de/uploads_file/attachment/attachment/439/1285.pdf).
